cpp/src/arrow/io/interfaces.h

```diff
@@ -200,7 +220,10 @@ class ARROW_EXPORT InputStream : virtual public FileInterface, virtual public Re
   InputStream() = default;
 };
 
-class ARROW_EXPORT RandomAccessFile : public InputStream, public Seekable {
+class ARROW_EXPORT RandomAccessFile
+    : public std::enable_shared_from_this<RandomAccessFile>,
```

Comment:
I'd be OK with having `enable_shared_from_this` on `FileInterface`. Of course any API that uses `shared_from_this` somewhere in its call stack will need to make the user aware that they need to have a `shared_ptr` control block in order to use it